### PR TITLE
Support pasting positions in the item property window, and refactor e…

### PR DESCRIPTION
**Pull request by Sickness-chan**
…xtracting positions from the clipboard

This will also improve pasting positions in the "Go to Position" window (CTRL +G)

Note: Pasting a position via the context menu in the item properties window is not supported, as the SpinCtrl doesn't emit a suitable event for pasting text, I have asked about it here: https://forums.wxwidgets.org/viewtopic.php?f=1&t=46035

### CODE REVIEW

**Comment by Mignari** [comment]
I got some errors
..\..\source\common.h(63): error C2065: 'Position' : undeclared identifier
..\..\source\common.cpp(224): error C3533: 'const auto &': a parameter cannot have a type that contains 'auto'
..\..\source\main_toolbar.cpp(492): error C2660: 'posFromClipboard' : function does not take 3 arguments

**Comment by Mignari** [comment]
It seems not good the g_gui depencency in this context. maybe its better to pass maxWidth and maxHeight as parameters

**Comment by Sickness-chan** [comment]
Sorry, I didn't have Visual Studio 2013 set up to build RME, let me know if there still are problems remaining.

File: source/positionctrl.cpp [hunk 1]
```diff
@@ -66,7 +68,8 @@ bool PositionCtrl::Enable(bool enable)
 void PositionCtrl::OnClipboardText(wxClipboardTextEvent& evt)
 {
 	Position position;
-	if(posFromClipboard(position.x, position.y, position.z)) {
+	Editor* editor = g_gui.GetCurrentEditor();
```
Comment (Mignari) [inline source/positionctrl.cpp]: This is another situation where g_gui's dependency is not good. PositionCtrl already have the maxx and maxy properties in contructor. You should use them

File: source/common.h [hunk 2]
```diff
@@ -60,7 +60,7 @@ std::wstring string2wstring(const std::string& utf8string);
 std::string wstring2string(const std::wstring& widestring);
 
 // Gets position values from ClipBoard
-bool posFromClipboard(int& x, int& y, int& z);
+bool posFromClipboard(Position& position, const int& mapWidth, const int& mapHeight);
```
Comment (Mignari) [inline source/common.h]: Like in PositionCtrl constructor, you can use int mapWidth= MAP_MAX_WIDTH, int mapHeight= MAP_MAX_HEIGHT to make it optional

File: source/common.cpp [hunk 3]
```diff
@@ -198,45 +200,66 @@ std::string wstring2string(const std::wstring& widestring)
...
+
+	static const std::regex exprs[] = {
+		// {x = 0, y = 0, z = 0}
+		std::regex(R"(\{x=(\d+),y=(\d+),z=(\d+)\})"),
+		// {"x": 0, "y": 0, "z": 0}
+		std::regex(R"(\{\"x\":(\d+),\"y\":(\d+),\"z\":(\d+)\})"),
+		// x, y, z
+		std::regex(R"((\d+),(\d+),(\d+))"),
+		// (x, y, z)
+		std::regex(R"(\((\d+),(\d+),(\d+)\))"),
```
[the remark this answers is not in the document]
Comment (Sickness-chan) [reply source/common.cpp]: There's no need for it, spaces are being stripped out before the input is checked against the regex. That is what this does
```cpp
input.erase(std::remove_if(input.begin(), input.end(), [](const auto &ch) -> bool {
	return std::isspace(ch);
}), input.end());```

[the remark this answers is not in the document]
Comment (gpedro) [reply source/common.cpp]: cool, so just ignore \s* on my regex

[the remark this answers is not in the document]
Comment (Sickness-chan) [reply source/common.cpp]: Having them seperate is cleaner IMO, but however you wish.

File: source/positionctrl.cpp (hunk 1 above)
Comment (Sickness-chan) [reply source/positionctrl.cpp]: Should be addressed with this commit https://github.com/hjnilsson/rme/pull/268/commits/b98c4f4c7e05ece1b177249275d8b4d7ce6690dc

File: source/common.h (hunk 2 above)
Comment (Sickness-chan) [reply source/common.h]: Should be addressed with this commit https://github.com/hjnilsson/rme/pull/268/commits/b98c4f4c7e05ece1b177249275d8b4d7ce6690dc

**Comment by gpedro** [comment]
LGTM

**Comment by Sickness-chan** [comment]
so is this gonna get merged or what

**Comment by Sickness-chan** [comment]
took too long

**Pull request closed without merging**